### ISSUE TITLE
Cypto Spot Trading bot


Der Trading-Bot überwacht kontinuierlich den Preis eines bestimmten Handelspaares auf der Bitmart-Börse. Er verwendet technische Indikatoren wie den RSI und den MACD, um Handelsentscheidungen zu treffen. Der Bot kauft, wenn der RSI niedrig ist und der MACD positiv ist, und setzt dabei Limit-Orders mit integrierten Stop-Loss- und Take-Profit-Niveaus, um Risiken zu minimieren und Gewinne zu sichern. Der Code ist darauf ausgelegt, in einem Endlosschleifenbetrieb zu laufen, wobei der Preis alle 60 Sekunden überprüft wird.

 

### ISSUE THREAD

**Issue by Atakan-24** [Atakan-24/TEST-2#1]
import requests
import json
import time
import numpy as np
from talib import RSI, MACD

# Bitmart API credentials
API_KEY = 'your_api_key'
API_SECRET = 'your_api_secret'

# Function to get latest price of a specific trading pair
def get_latest_price(symbol):
    url = f'https://api-cloud.bitmart.com/spot/v1/ticker?symbol={symbol}'
    response = requests.get(url)
    data = response.json()
    return float(data['data']['ticker']['last'])

# Function to place a limit buy order with stop-loss and take-profit
def place_limit_buy_order(symbol, quantity, price, stop_loss=None, take_profit=None):
    endpoint = 'https://api-cloud.bitmart.com/spot/v1/submit_order'
    payload = {
        'symbol': symbol,
        'type': '2',  # Limit buy
        'quantity': quantity,
        'price': price,
        'side': 'buy',
        'timeInForce': 'GTC',  # Good 'Til Canceled
        'recvWindow': 5000,
        'timestamp': int(time.time() * 1000)
    }
    if stop_loss:
        payload['stopPrice'] = stop_loss
    if take_profit:
        payload['stopPrice'] = take_profit
    payload['sign'] = generate_signature(payload)
    response = requests.post(endpoint, data=payload)
    return response.json()

# Function to place a limit sell order with stop-loss and take-profit
def place_limit_sell_order(symbol, quantity, price, stop_loss=None, take_profit=None):
    endpoint = 'https://api-cloud.bitmart.com/spot/v1/submit_order'
    payload = {
        'symbol': symbol,
        'type': '2',  # Limit sell
        'quantity': quantity,
        'price': price,
        'side': 'sell',
        'timeInForce': 'GTC',  # Good 'Til Canceled
        'recvWindow': 5000,
        'timestamp': int(time.time() * 1000)
    }
    if stop_loss:
        payload['stopPrice'] = stop_loss
    if take_profit:
        payload['stopPrice'] = take_profit
    payload['sign'] = generate_signature(payload)
    response = requests.post(endpoint, data=payload)
    return response.json()

# Function to generate signature for authentication
def generate_signature(payload):
    # Implement your signature generation logic here
    return 'your_signature'

# Main function to run the trading bot
def main():
    symbol = 'ushark_usdt'  # Replace with your desired trading pair
    quantity = 10  # Replace with your desired quantity

    while True:
        price = get_latest_price(symbol)
        print(f'Latest price of {symbol}: {price}')

        # Calculate RSI
        rsi_period = 14
        rsi_values = RSI(np.array([price]), timeperiod=rsi_period)
        rsi = rsi_values[0][-1]

        # Calculate MACD
        macd, signal, _ = MACD(np.array([price]), fastperiod=12, slowperiod=26, signalperiod=9)
        macd_diff = macd[-1] - signal[-1]

        print(f'RSI: {rsi}, MACD Difference: {macd_diff}')

        # Advanced risk management: Implementing stop-loss and take-profit orders
        stop_loss = price * 0.95  # 5% stop loss below entry price
        take_profit = price * 1.1  # 10% take profit above entry price

        # Buy if RSI is low and MACD difference is positive
        if rsi <= 30 and macd_diff > 0:
            print('Low RSI and positive MACD difference. Placing buy order...')
            buy_price = price * 0.99  # Buy slightly below current price
            buy_response = place_limit_buy_order(symbol, quantity, buy_price, stop_loss, take_profit)
            print('Buy order response:', buy_response)

        time.sleep(60)  # Check price every 60 seconds

if __name__ == '__main__':
    main()

Ist der Code gut?
